test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,293 @@
+    arg_dict_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'arg_dict')
+    script_module.save(module_file_path)
+    serialize_arg_dict_as_script_module(test_params.arg_dict).save(arg_dict_file_path)
+
+    cpp_test_name = '{}_{}'.format(test_params.module_variant_name, 'test_forward_backward')
+    cpp_test_fn = getattr(unit_test_class.module_impl_check_cpp_module, cpp_test_name)
+
+    def run_cpp_test_fn_and_check_output():
+        forward_output_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'forward_output')
+        backward_grad_dict_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'backward_grad_dict')
+
+        cpp_test_fn(arg_dict_file_path, module_file_path, forward_output_file_path, backward_grad_dict_file_path)
+        cpp_output = torch.load(forward_output_file_path)
+        cpp_grad_dict = torch.load(backward_grad_dict_file_path)
+
+        # Check that forward outputs are equal
+        unit_test_class.assertTrue(
+            torch.allclose(python_output, cpp_output),
+            generate_error_msg("forward output", cpp_output, python_output))
+
+        # Check that module parameter gradients are equal after backward pass
+        unit_test_class.assertEqual(
+            len(python_grad_dict), len(cpp_grad_dict),
+            generate_error_msg("# of parameters", len(cpp_grad_dict), len(python_grad_dict)))
+        for key in python_grad_dict:
+            unit_test_class.assertTrue(
+                key in cpp_grad_dict,
+                generate_error_msg("\"Does module have a parameter named `{}`?\"".format(key[:-5]), False, True))
+            unit_test_class.assertTrue(
+                torch.allclose(python_grad_dict[key], cpp_grad_dict[key]),
+                generate_error_msg("gradient of `{}`".format(key[:-5]), cpp_grad_dict[key], python_grad_dict[key]))
+
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs
+    assert os.path.exists(cpp_tmp_folder)
+    shutil.rmtree(cpp_tmp_folder)
+
+def test_torch_nn_module_variant(unit_test_class, test_params):
+    test_forward_backward(unit_test_class, test_params)
+
+def compute_module_name(test_params_dict):
+    fullname = test_params_dict.get('fullname', None)
+    if fullname:
+        module_name = fullname.split('_')[0]
+    else:
+        module_name = test_params_dict.get('module_name')
+    return module_name
+
+def process_test_params_for_module(test_params_dict, device, test_instance_class):
+    module_name = compute_module_name(test_params_dict)
+    test_params_dict['constructor'] = test_params_dict.get('constructor', getattr(torch.nn, module_name))
+    test_instance = test_instance_class(**test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    module_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
```

Comment:
I see multiple places that we concatenate multiple strings to form a variable name, sometimes use hardcoded stuff as well. Can we put a comment here or somewhere to give a full name sample or example so it won't be hard to trace what is this name look like.

Reply:
Thanks for the catch! I added example outputs for `module_variant_name` and `functional_variant_name`. Please let me know if other examples are needed and I can add them in as well :D

Reply:
thx Will